### PR TITLE
fix(presenter): resize desktop track to 720p when presenter starts

### PR DESCRIPTION
**Pull request by jallamsetty1**
Resize the desktop MediaStreamTrack to 720p using applyConstraints for the following cases
- When the track height cannot be determined (firefox doesn't seem to return the track settings.
- Determine if the window is in portrait/landscape mode and resize only the dimension which is bigger than 1280/720. The other side resizes to preserve the aspect ratio.
- Add more 4:3 resolutions to pick from when the camera stream is obtained.
When the presenter is turned on, we start sending full 30 fps which can be very cpu intensive if the shared desktop window is very big.

### CODE REVIEW

**Comment by saghul** [comment]
Question: does this code behave properly when sharing a window in "portrait" mode? That is, when the height is larger than the width. Looks like we may needdlessly resize it, unless the bworser is normalizing those values, regardless of orientation, which doesn't seem likely.

**Comment by jallamsetty1** [comment]
> Question: does this code behave properly when sharing a window in "portrait" mode? That is, when the height is larger than the width. Looks like we may needdlessly resize it, unless the bworser is normalizing those values, regardless of orientation, which doesn't seem likely.

@saghul , I agree. We would be resizing a window that is in portrait mode. However, I feel the presenter mode looks better when we resize it. I have attached screenshots w/ and w/o resizing. We can think of an alternative if this isn't acceptable.
Screen Shot 2020-02-05 at 9 29 05 AM
Screen Shot 2020-02-05 at 9 29 05 AM

**Comment by saghul** [comment]
Can you trry sharing an iOS simulator for example? I think a text editor is not a great example.

**Comment by jallamsetty1** [comment]
I see what you are saying. So just detect if its in portrait mode (height > width) and not resize if that is the case ?

**Comment by saghul** [comment]
> I see what you are saying. So just detect if its in portrait mode (height > width) and not resize if that is the case ?

I think we need to detect portrait mode and resize the appropriate dimension, height or width.

File: conference.js [hunk 1]
```diff
@@ -1664,27 +1664,47 @@ export default {
...
+        // Create a new presenter track and apply the presenter effect.
         if (!this.localPresenterVideo && !mute) {
-            // create a new presenter track and apply the presenter effect.
             let { height } = this.localVideo.track.getSettings();
-
-            // Workaround for Firefox since it doesn't return the correct width/height of the desktop stream
-            // that is being currently shared.
-            if (!height) {
-                const desktopResizeConstraints = {
+            const { aspectRatio, width } = this.localVideo.track.getSettings();
```
Comment (saghul) [inline conference.js]: you can import height here too, since it's not reassigned and can be a const

File: conference.js [hunk 2]
```diff
@@ -1664,27 +1664,47 @@ export default {
...
+            let resizeDesktopStream = false;
+
+            // Determine the constraints if the desktop track needs to be resized.
+            // Resizing is needed when the resolution cannot be determined or when
+            // the window is bigger than 1280x720.
+            if (height && width && aspectRatio) {
+                const advancedConstraints = [ { aspectRatio } ];
+                const resizeHeight = height >= width;
+
+                // Determine which dimension needs resizing and resize only that side
```
Comment (saghul) [inline conference.js]: The logic here doesn't look correct. If you have a portrait 720p screen it means it's 720x1280.  Now imagine it's a tad bigger, say 800x1280. The code below will resize the height tto 720, but that's not what you want, is it?  What we want is to resize the width to be 720.

```
// Somewhere else:
const DESKTOP_STREAM_CAP = 720;

const isPortrait = height >= width;
if (isPortrait && width > DESKTOP_STREAM_CAP) {
    resizeDesktopStream = true;
    advancedConstraints.push({ width: DESKTOP_STREAM_CAP });
} else if (!isPortrait && height > DESKTOP_STREAM_CAP) {
    resizeDesktopStream = true;
    advancedConstraints.push({ height: DESKTOP_STREAM_CAP });
}
```

Isn't the above what we're after?

File: react/features/base/tracks/functions.js [hunk 3]
```diff
@@ -27,7 +27,7 @@ export async function createLocalPresenterTrack(options, desktopHeight) {
 
     // compute the constraints of the camera track based on the resolution
     // of the desktop screen that is being shared.
-    const cameraHeights = [ 180, 270, 360, 540, 720 ];
+    const cameraHeights = [ 120, 180, 240, 360, 480, 600, 720 ];
```
Comment (saghul) [inline react/features/base/tracks/functions.js]: why is this change necessary?

File: conference.js [hunk 4]
```diff
@@ -1664,27 +1664,47 @@ export default {
...
-            if (!height) {
-                const desktopResizeConstraints = {
+            const { aspectRatio, width } = this.localVideo.track.getSettings();
+            let desktopResizeConstraints = {};
+            let resizeDesktopStream = false;
+
+            // Determine the constraints if the desktop track needs to be resized.
+            // Resizing is needed when the resolution cannot be determined or when
+            // the window is bigger than 1280x720.
+            if (height && width && aspectRatio) {
```
Comment (saghul) [inline conference.js]: Is there a device which doesn't expose the aspecct ratio? I guess we should be able to figure it out if we have height and width, right?

File: conference.js (hunk 2 above)
Comment (jallamsetty1) [reply conference.js]: Since we are using height for determining the resolution of the camera, I was making sure it is not greater than 720 but I see your point.

File: conference.js (hunk 1 above)
Comment (jallamsetty1) [reply conference.js]: height is getting reassigned at line # 1707.

File: conference.js (hunk 4 above)
Comment (jallamsetty1) [reply conference.js]: yes we can calculate aspect ratio from height and width, will remove the check for aspect ratio here.

File: react/features/base/tracks/functions.js (hunk 3 above)
Comment (jallamsetty1) [reply react/features/base/tracks/functions.js]: When testing with the iOS simulator I realized that we need lower resolutions for the camera, so I added more 4:3 resolutions.
